File: src/linkerz_crawler_core/src/main/java/org/linkerz/crawler/core/job/CrawlJob.java
/*
 * Copyright (C) 2012 - 2013 LinkerZ
 */

package org.linkerz.crawler.core.job;

import org.linkerz.core.job.AbstractJob;
import org.linkerz.crawler.core.job.config.CrawlJobConfig;
import org.linkerz.crawler.core.model.WebLink;

/**
 * The Class CrawlJob.
 *
 * @author Nguyen Duc Dung
 * @since 7/3/12, 3:12 AM
 */
public class CrawlJob extends AbstractJob<Void, CrawlJobConfig> {

    private WebLink webLink;
    private CrawlJobConfig config;

    public CrawlJob() {
        //Only for Serializable.
    }

    public CrawlJob(String url) {
        this(url, new CrawlJobConfig());
    }

    public CrawlJob(String url, CrawlJobConfig config) {
        this.webLink = new WebLink(url);
        this.config = config;
    }

    public CrawlJob(WebLink webLink, CrawlJobConfig config) {
        this.webLink = webLink;
        this.config = config;
    }

    public CrawlJob(WebLink webLink) {
        this(webLink, new CrawlJobConfig());
    }

    public WebLink getWebLink() {
        return webLink;
    }

    public void setWebLink(WebLink webLink) {
        this.webLink = webLink;
    }


    @Override
    public void setConfig(CrawlJobConfig config) {
        this.config = config;
    }

    @Override
    public CrawlJobConfig getConfig() {
        return config;
    }
}

File: src/linkerz_crawler_core/src/main/scala/org/linkerz/crawler/core/job/CrawlJob.scala
/*
 * Copyright (C) 2012 - 2013 LinkerZ
 */

package org.linkerz.crawler.core.job

import org.linkerz.job.queue.core.Job
import org.linkerz.crawler.core.model.{WebPage, WebUrl}
import collection.mutable.ListBuffer
import scala.Some

/**
 * The Class CrawlJob.
 *
 * @author Nguyen Duc Dung
 * @since 7/29/12, 12:55 AM
 *
 */

case class CrawlJob(webUrl: WebUrl) extends Job {

  private var _parent: Option[CrawlJob] = None
  private var _result: Option[WebPage] = None

  /**
   * The depth of the job from the first job.
   */
  private var _depth: Int = 0

  var code: CrawlJob.Status = CrawlJob.DONE

  /**
   * Max depth for a crawl job.
   */
  var maxDepth: Int = 1

  var onlyCrawlInSameDomain: Boolean = true

  /**
   * Only crawl the url match with this regex.
    */
  var urlRegex: List[String] = _

  /**
   * For those of url match with this regex will not be crawl.
   */
  var excludeUrl: List[String] = _

  /**
   * String url.
   * @param url
   */
  def this(url: String) {
    this(new WebUrl(url))
  }

  /**
   *
   * @param webUrl
   * @param parentJob
   */
  def this(webUrl: WebUrl, parentJob: CrawlJob) {
    this(webUrl)
    assert(parent != null)
    this.parent = Some(parentJob)
  }

  def result = {
    _result
  }

  def result_=(result: Option[WebPage]) {
    _result = result
  }

  override def parent = _parent

  def parent_=(parent: Option[CrawlJob]) {
    _parent = parent
  }

  def depth: Int = {
    if (!parent.isEmpty) {
      _depth = parent.get.depth + 1
    }
    _depth
  }
}

object CrawlJob extends Enumeration {
  type Status = Value
  val DONE, SKIP, ERROR = Value
}